kuma/wiki/templatetags/ssr.py

```diff
@@ -0,0 +1,123 @@
+import json
+import subprocess
+import sys
+import time
+
+from django_jinja import library
+
+# A global that holds the reference to a Node child process that does
+# server-side rendering. This will be initialized when first
+# needed. But because of startup overhead we don't want to create this
+# process each time, so we need to keep a referece here so that
+# multiple calls to server_side_render() can use it.
+renderer = None
+
+# The output of the {{render_react_app(data)}} is going to look
+# like this, with optional HTML and state replacing the `{}` markers.
+OUTPUT_TEMPLATE = (
+    u'<div id="react-container">{}</div>\n'
+    u'<script>window._document_data = {};</script>\n'
+)
+
+
+@library.global_function
+def render_react_app(data, ssr=True):
+    """
+    Render a script tag to define the data and any other HTML tags needed
+    to enable the display of a React-based UI. By default, this does
+    server side rendering, falling back to client-side rendering if
+    the SSR attempt fails. Pass False as the second argument to do
+    client-side rendering unconditionally.
+
+    Note that we are not defining a generic Jinja template tag here.
+    The code in this file is specific to Kuma's React-based UI.
+    """
+    if ssr:
+        return server_side_render(data)
+    else:
+        return client_side_render(data)
+
+
+def client_side_render(data):
+    """
+    Output an empty <div> and a script with complete state so that
+    the UI can be rendered on the client-side.
+    """
+    # Convert to JSON and be sure the string "</script>" does not appear
+    serializedState = json.dumps(data).replace('</', '<\\/')
+    return OUTPUT_TEMPLATE.format('', serializedState)
+
+
+def server_side_render(data):
+    """
+    Pre-render the React UI to HTML and output it in a <div>, and then
+    also pass the necessary serialized state in a <script> so that
+    React on the client side can sync itself with the pre-rendred HTML.
+
+    If any exceptions are thrown during the server-side rendering, we
+    fall back to client-side rendering instead.
+    """
+
+    global renderer
+    start = time.time()
+
+    # Try server side rendering
+    try:
+        # If we haven't started a renderer process, or if that process
+        # exited for some reason, then start one.
+        if not renderer or renderer.poll() is not None:
+            print "Starting renderer subprocess"
+            renderer = subprocess.Popen(['/usr/bin/env',
+                                         'node',
+                                         '/app/kuma/javascript/dist/ssr.js'],
+                                        stdin=subprocess.PIPE,
+                                        stdout=subprocess.PIPE,
+                                        stderr=sys.stderr)
```

Comment:
I'm working on a test for ssr.py, and realize that I'm going to move the code about this into a separate get_renderer() function, so that I can mock it in the tests.